onnxruntime/core/graph/graph.cc

```diff
@@ -4046,6 +4047,229 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
+  } else {
+    unique_id.append(else_branch);
+  }
+
+  unique_id = GenerateNodeName(unique_id);
+
+  auto make_unique = [&unique_id](const std::string& name) {
+    return unique_id + '_' + name;
+  };
+
+  // Check if the name is an input or implicit input.
+  // These are not renamed.
+  // Implicit inputs would cover both If node input and implicit inputs.
+  // Reason: there are no explicit inputs to the subgraphs, and the subgraph's
+  // implicit inputs must be covered by the implicit inputs of the If node.
+  InlinedHashSet<std::string_view> outer_scope_values;
+  const auto if_implicit_inputs = if_node.ImplicitInputDefs();
+  outer_scope_values.reserve(if_implicit_inputs.size());
+
+  for (const auto* input : if_implicit_inputs) {
+    const auto& name = input->Name();
+    ORT_IGNORE_RETURN_VALUE(outer_scope_values.emplace(name));
+  }
+
+  // Name mapping from the graph to inline to the graph we are inlining into
+  // we also use this to process any subgraphs in the graph we are inlining
+  InlinedHashMap<std::string, std::string> name_mapping;
+
+  // We are going to map the outputs of the graph to inline to the outputs of the If node.
+  // They are assumed to be in the same order.
+  const auto node_output_defs = if_node.OutputDefs();
+  const auto graph_output_defs = graph_to_inline.GetOutputs();
+  for (size_t i = 0; i < graph_output_defs.size(); ++i) {
+    name_mapping.emplace(graph_output_defs[i]->Name(), node_output_defs[i]->Name());
+  }
+
+  // we would like to move the entries that can be potentially big
+  for (int i = 0, limit = graph_to_inline.graph_proto_->initializer_size(); i < limit; ++i) {
+    auto* initializer = graph_to_inline.graph_proto_->mutable_initializer(i);
+    const std::string src_name = initializer->name();
+
+#if !defined(DISABLE_SPARSE_TENSORS)
+    bool has_sparse_origin = false;
+    if (!graph_to_inline.sparse_tensor_names_.empty()) {
+      auto hit = graph_to_inline.sparse_tensor_names_.find(src_name);
+      if (hit != graph_to_inline.sparse_tensor_names_.cend()) {
+        has_sparse_origin = true;
+        // Erase the entry that will be invalidated
+        graph_to_inline.sparse_tensor_names_.erase(hit);
+      }
+    }
+#endif
+
+    graph_to_inline.name_to_initial_tensor_.erase(src_name);
+    const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
+    *tensor = std::move(*initializer);
+
+    // Check if this is an output of the graph
+    auto hit = name_mapping.find(src_name);
+    if (hit != name_mapping.cend()) {
+      tensor->set_name(hit->second);
+    } else {
+      auto new_name = GenerateNodeArgName(make_unique(src_name));
+      ORT_IGNORE_RETURN_VALUE(name_mapping.emplace(src_name, new_name));
+      tensor->set_name(std::move(new_name));
+    }
+
+    auto insert_result = name_to_initial_tensor_.emplace(tensor->name(), tensor);
+    ORT_ENFORCE(insert_result.second, "Initializer name: ", tensor->name(), " from graph: ",
```

Comment:
It seems to me that we could have a name-conflict here that has nothing to do with specializing code (which, I guess, is involved only with function-inlining). Specifically, what happens if we inline two if-statements one after another? And if the inlined-branches in the two if-statements have initializers with the same name? They don't conflict with each other before if-branch-folding. First if-branch-folding should be fine. But the second one might produce a name conflict, right?

Reply:
My fault, I missed the renaming that happens just a few lines above! I was misled by the error message "Check specializing code.", which may be misleading. I guess as long as `make_unqiue` above does its job, this shouldn't fail.

Reply:
I have adjusted the error message.